api/app/jobs/viirs_snow.py

```diff
@@ -1,11 +1,222 @@
+from datetime import date, timedelta
+from osgeo import gdal
+import asyncio
+import glob
 import logging
-
-from app import configure_logging
+import os
+import requests
+import shutil
+import sys
+import tempfile
+from app import config, configure_logging
+from app.db.crud.snow import get_last_processed_snow_by_source, save_processed_snow
+from app.db.database import get_async_read_session_scope, get_async_write_session_scope
+from app.db.models.snow import ProcessedSnow, SnowSourceEnum
+from app.rocketchat_notifications import send_rocketchat_notification
+from app.utils.s3 import get_client
 
 logger = logging.getLogger(__name__)
 
+BC_BOUNDING_BOX = "-139.06,48.3,-114.03,60"
+NSIDC_URL = "https://n5eil02u.ecs.nsidc.org/egi/request"
+PRODUCT_VERSION = 2
+SHORT_NAME = "VNP10A1F"
+LAYER_VARIABLE = "/VIIRS_Grid_IMG_2D/CGF_NDSI_Snow_Cover"
+RAW_SNOW_COVERAGE_NAME = 'raw_snow_coverage.tif'
+RAW_SNOW_COVERAGE_CLIPPED_NAME = 'raw_snow_coverage_clipped.tif'
+
+
+class ViirsSnowJob():
+    """ Job that downloads and processed VIIRS snow coverage data from the NSIDC (https://nsidc.org).
+    """
+
+    async def _get_last_processed_date(self) -> date:
+        """ Return the date of the most recent successful processing of VIIRS snow data.
+
+        :return: The date of the most recent successful processing of VIIRS snow data.
+        :rtype: date
+        """
+        async with get_async_read_session_scope() as session:
+            last_processed = await get_last_processed_snow_by_source(session, SnowSourceEnum.viirs)
+            return None if last_processed is None else last_processed[0].for_date.date()
+    
+
+    def _download_viirs_granules_by_date(self, for_date: date, path: str, file_name: str):
+        """ Download VIIRS snow data for the specified date.
+        
+        :param for_date: The date of interest.
+        :type for_date: date
+        :param path: The path to a temporary directory to download the data to.
+        :type path: str
+        :param file_name: The name to assign to the downloaded file/archive.
+        :type: file_name: str
+        """
+        # Interesting flow required by the NSIDC API. We need to issue a request without credentials which will fail
+        # with a 401. The URL in the response object is the authentication endpoint, so we issue a request to it with basic auth.
+        # Now we are authenticated with a session cookie and can make additional requests to the API.
+        # Furthermore, it looks like we're stuck with synchronous requests.
+        logger.info(f"Downloading VIIRS snow coverage data for date: {for_date.strftime('%Y-%m-%d')}")
+        session = requests.session()
+        resp = session.get(f"https://n5eil02u.ecs.nsidc.org/egi/capabilities/{SHORT_NAME}.{PRODUCT_VERSION}.xml")
+        session.get(resp.url, auth=(config.get("NASA_EARTHDATA_USER"),config.get("NASA_EARTHDATA_PWD")))
+        # Check if request was successful
```

Comment:
Looks like we do 2 requests here? One to authenticate, then one to get the data. Where are we checking that the auth request was successful?

Reply:
Oops, that's a leftover comment at this spot that I'll remove. The check to see if we've made a successful request actually occurs below with `request.raise_for_status()`. If the request for data returns anything but a 200 this call with raise an error which we subsequently handle.